runtime/vm/mz64/UpcallThunkGen.cpp

```diff
@@ -56,179 +57,433 @@ extern "C" {
-                }
-                fprArgCount++;
-                break;
-            default:
-                Assert_VM_unreachable();
-        }
-        // On z/OS the first four double/float type arguments are passed in FPRs, and the first three arguments of any other type
-        // are passed in GPRs 1-3. If we have passed those limits while iterating through the argument list, then we can stop
-        // counting the number of instructions. We won't need to store the remaining parameters because they are already
-        // in memory.
-        if (fprArgCount>=4 && gprArgCount>=3)
-            break;
+	// Loop through the arguments and count how many instructions you'll need to copy data from registers to memory, and how many
+	// bytes of space you'll need to allocate in thunk and new stack frame to store the instructions + parameter info.
+	for (I_32 i = 0; i < lastSigIdx; i++) {
+		U_32 argSizeInBytes = sigArray[i].sizeInByte;
+		switch (sigArray[i].type) {
+		case J9_FFI_UPCALL_SIG_TYPE_CHAR:    /* Fall through */
+		case J9_FFI_UPCALL_SIG_TYPE_SHORT:   /* Fall through */
+		case J9_FFI_UPCALL_SIG_TYPE_INT32:   /* Fall through */
+		case J9_FFI_UPCALL_SIG_TYPE_POINTER: /* Fall through */
+		case J9_FFI_UPCALL_SIG_TYPE_INT64:
+			if (gprArgCount < maxGPRArgCount) {
+				numInstructions++;
+				gprArgCount++;
+			}
+			break;
+		case J9_FFI_UPCALL_SIG_TYPE_FLOAT:
+		case J9_FFI_UPCALL_SIG_TYPE_DOUBLE:
+			if (fprArgCount < maxFPRArgCount) {
+				numInstructions++;
+				fprArgCount++;
+			}
+			if (i < 3) {
+				gprArgCount++; // If one of the first three doublewords of the argument area is float/double, then the corresponding GPR is undefined. So we must ignore any value in there.
+			}
+			break;
+		case J9_FFI_UPCALL_SIG_TYPE_STRUCT_AGGREGATE_ALL_SP:
+			if (argSizeInBytes == 2 * sizeof(float)) {
+				// a pure float struct with only 2 floats is passed in 2 FPRs.
+				// If only one FPR is free, then the first half of the struct is
+				// in FPR6, and the second half is in the corresponding slot
+				// in the argument area, right adjusted.
+				if (fprArgCount <= maxFPRArgCount - 2) {
+					numInstructions += 2;
+					fprArgCount += 2;
+				}
+				else if (fprArgCount < maxFPRArgCount) {
+					numInstructions++;
+					fprArgCount++;
+				}
+			}
+			else { // All other pure float structs are passed in GPR (if there's space)
+				if (gprArgCount < maxGPRArgCount) {
+					// gprsNeededForEntireStruct below represents the number of GPRs we would need hypothetically to load the entire struct in GPRs.
+					// Since this is a pure float structure, the size will always be a multiple of 4. So if we divide by 8, like below, the result
+					// will either be some whole number 'X', or 'X.5'. For `X.5`, adding 0.5 to the sum will always round UP to the nearest whole number.
+					int gprsNeededForEntireStruct = (int)(((float)argSizeInBytes / 8.0F) + 0.5);
```

Comment:
Isn't this just aligning the size to double word boundary ?  Can't we do something like following,
```
((argSizeInBytes-1)/8)+1
```

Reply:
This, along with your other similar comments, should all be taken care of now in: b6cac56c59ac275b1cd08ba52da968d2a7cb5e94.